From 0c8986e633787bcd25f4e80cd67bb8059ea6b74e Mon Sep 17 00:00:00 2001
From: "Dr. Greg Wettstein" <greg@enjellic.com>
Date: Fri, 24 Feb 2017 17:57:06 -0600
Subject: [PATCH 09/24] Add content digest of subject to subject identity.

This update adds support for including the digest value of the
subject into the aggregate subject identity.  This simply
involves creating an array within the subject_identity structure
and copying of the digest value from the event into that
structure.
---
 security/integrity/ima/ima_identity.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/security/integrity/ima/ima_identity.c b/security/integrity/ima/ima_identity.c
index 2b7145322f2b..7f9fba3e603d 100644
--- a/security/integrity/ima/ima_identity.c
+++ b/security/integrity/ima/ima_identity.c
@@ -47,6 +47,8 @@ struct subject_identity {
 
 	u32 name_length;
 	char name[WP256_DIGEST_SIZE];
+
+	char digest[WP256_DIGEST_SIZE];
 } __attribute__((__packed__));
 
 /*
@@ -845,8 +847,9 @@ int ima_get_subject(struct ima_event_data *event_data, char *subject)
 		return PTR_ERR(tfm);
 
 	filename = event_data->filename;
-
 	memset(&identity, '\0', sizeof(struct subject_identity));
+
+	/* Add subject characteristics. */
 	if (event_data->file) {
 		inode = file_inode(event_data->file);
 		identity.uid = from_kuid(&init_user_ns, inode->i_uid);
@@ -854,6 +857,10 @@ int ima_get_subject(struct ima_event_data *event_data, char *subject)
 		identity.mode = inode->i_mode;
 	}
 
+	/* Add the digest of the subject. */
+	memcpy(identity.digest, event_data->iint->ima_hash->digest,
+	       sizeof(identity.digest));
+
 	retn = ima_get_subject_identity(tfm, identity, subject);
 	pr_info("[%s]: mode=o%o, uid=%u, gid=%u, name=%s, subject=%*phN\n",
 		__func__, identity.mode, identity.uid, identity.gid,
-- 
2.11.0

